ENH: Propose standard policy for dropping support of old Python versions (#14086)
numpy/numpy: doc/neps/nep-0029-deprecation_policy.rst

:Author: Thomas A Caswell <tcaswell@gmail.com>, Andreas Mueller, Brian Granger, Madicken Munk, Ralf Gommers, Matt Haberland <mhaberla@calpoly.edu>, Matthias Bussonnier <bussonniermatthias@gmail.com>, Stefan van der Walt
:Status: Draft
:Type: Informational Track
:Created: 2019-07-13

Abstract
--------

All projects across the ecosystem should adopt a common time window
based policy for increasing the minimum version of Python and NumPy
that downstream projects support.  By standardizing this policy
across the community we will make it easier for downstream projects to
plan.

This is an unusual NEP in that it is recommendations for community-wide
policy and not for changes to NumPy itself.  However, we do not
currently have a place for SPEEPs (Scientific Python Ecosystem
Enhancement Proposals) and given NumPy's central role in our community
this seems like the proper place to document this.

This is being put forward by maintainers of Matplotlib, scikit-learn,
IPython, Jupyter, yt, SciPy, NumPy, and scikit-image.

Detailed description
--------------------

When making a new major or minor release, projects should support all

> minor versions of ``Python`` initially released in the prior 42

Reviewer: How does this support algorithm handle the release of 4.0.0

Author: I think if we get kernel-style major version bumps, we can treat it as any other minor release and I don't think there is upstream appetite for a 2->3 style 3->4 transition.  Not sure how to phrase that succinctly though.

Reviewer 2: I suggest to just put a disclaimer in about major versions, since it's not really possible to predict in advance what the right strategy is. It really depends on what breaks in 4.0 and how the community reacts to that.

Author: Added a sentence at the end of this section.

the month after ``Python`` and would effectively only support one
version of ``Python`` that has an install base (and then the window
is 42).  This buffer also makes us resilient to minor fluctuations /
delays in the ``Python`` release schedule.

Because the threshold for dropping a version of ``Python`` is based
on historical release dates and a project's plans, the decision to drop
support for a given version of ``Python`` can be made very early in
the release process.

There will be some unavoidable mismatch in supported version of
``Python`` between packages immediately after a version of
``Python`` ages out.  However this should not last longer than one
release cycle of each of the projects, and when a given project
does a minor or major release, it is guaranteed that there will be a
stable release of all other projects that support the set of
``Python`` the new release will support.

Implementation
--------------

We suggest that all projects adopt the following language into their
development guidelines:

   - support minor versions of ``Python`` initially released
     42 months prior to our planned release date

>    - always support at least the 2 latest versions of ``Python``

Reviewer 3: should this say minor versions?

   - support minor versions of ``numpy`` initially released in the 24
     months prior to our planned release date or oldest that supports the
     minimum Python version (whichever is higher)

   The minimum supported version of ``Python`` will be set to
   ``python_requires`` in ``setup`` and all supported versions of
   Python will be in the test matrix and have binary artifacts built
   for releases.

   We will bump the minimum Python and NumPy versions as we can on
   every minor and major release, but never on a patch release.

For other dependencies, adopt similar time windows of 24 months or
shorter.

Backward compatibility
----------------------

No issues other than the intentional dropping of old version of
``Python`` and dependencies.

Alternatives
------------

Ad-Hoc
~~~~~~

Projects could on every release evaluate if they want to increase
the minimum version of Python supported.  While this is a notionally
simple policy, it makes it hard for downstream users to predict what
the future minimum versions will be.  As there is no objective threshold
to when the minimum version should be dropped, it is easy for these

> discussions to devolve into bike shedding and acrimony.

Reviewer: ```suggestion
discussions to devolve into [bike shedding](https://en.wikipedia.org/wiki/Wikipedia:Avoid_Parkinson%27s_bicycle-shed_effect) and acrimony.
```

All Python Software Foundation supported versions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This is a very clear and conservative approach.  However, it means that
there is 4 year lag between when new language features come into the
language and when the projects are able to use them.  Additionally,
for projects that have a significant component of compiled extensions
this requires building many binary artifacts for each release.

For the case of NumPy, many projects carry workarounds to bugs that
are fixed in subsequent versions of NumPy.  Being proactive about
increasing the minimum version of NumPy will allow downstream
packages to carry fewer version-specific patches.

Default version on Linux distribution
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The policy could be support the version of Python that ships by

Reviewer: ```suggestion
The policy could be to support the version of Python that ships by
```

==========================================================================================
NEP 29 — A standard community policy for dropping support of old Python and NumPy versions
==========================================================================================

:Author: Thomas A Caswell <tcaswell@gmail.com>, Andreas Mueller, Brian Granger, Madicken Munk, Ralf Gommers, Matt Haberland <mhaberla@calpoly.edu>, Matthias Bussonnier <bussonniermatthias@gmail.com>, Stefan van der Walt
:Status: Draft
:Type: Informational Track
:Created: 2019-07-13

Abstract
--------

All projects across the ecosystem should adopt a common time window
based policy for increasing the minimum version of Python and NumPy
that downstream projects support.  By standardizing this policy
across the community we will make it easier for downstream projects to
plan.

This is an unusual NEP in that it is recommendations for community-wide

> policy and not for changes to NumPy itself.  However, we do not

Author: Did this "by hand" to work around the lack of multi-line diffs

currently have a place for SPEEPs (Scientific Python Ecosystem
Enhancement Proposals) and given NumPy's central role in our community
this seems like the proper place to document this.

This is being put forward by maintainers of Matplotlib, scikit-learn,
IPython, Jupyter, yt, SciPy, NumPy, and scikit-image.

Detailed description
--------------------

When making a new major or minor release, projects should support all
minor versions of ``Python`` initially released in the prior 42
months from their anticipated release date and all minor version of
NumPy released in the prior 24 months.

The diagram::

       Jan 16      Jan 17      Jan 18      Jan 19      Jan 20
       |           |           |           |           |
  +++++|+++++++++++|+++++++++++|+++++++++++|+++++++++++|++++++++++++
   |              |                  |               |
   py 3.5.0       py 3.6.0           py 3.7.0        py 3.8.0
  |-----------------------------------------> Feb19

>             |-----------------------------------------> Dec19

Author: The right facing arrow was meant to indicate that this is a semi-infinite support window and packages should strive to support future Python versions on a given release series.

>             |-----------------------------------------> Dec19

Reviewer 4: Fair enough. Thanks for clarifying.

>             |-----------------------------------------> Dec19

Author: How could this be made clearer? would

```
|-----------------------------------------|-- Dec19 --->
```

be better?

I like my ascii art, but maybe the tables are clearer and we should drop this altogether?

>             |-----------------------------------------> Dec19

Reviewer 5: A list of release dates would probably be easier to maintain. It can be a PITA to always google for the release dates...

>             |-----------------------------------------> Dec19

Author: I don't think this should be updated over time, I think _seeing_ the windows help with understanding.

                      |-----------------------------------------> Nov20

shows the support windows for ``Python``.  A project with a major or
minor release in Feb19 should support py35 and newer, a project with a
major or minor release in Dec19 should support py36 and newer, and a
project with a major or minor release in Nov20 should support py37 and
newer.

The current Python release cadence is 18 months so a 42 month window
ensures that there will always be at least two versions of ``Python``
in the window.  By padding the window by 6 months from the anticipated
``Python`` cadence we avoid the edge cases where a project releases

> the month after ``Python`` and would effectively only support one

Reviewer 5: Why not just "last three minor Python releases", where the usual version numbering is `major:minor:micro`

> the month after ``Python`` and would effectively only support one

Reviewer 6: Note that there's a terminology conflict here between CPython and basically everywhere else: 3.6, 3,7, etc are frequently referred to as major releases, and 3.x.y are minor releases. So if you say "minor Python releases", people *are* going to get confused.

One way to get around that would be to refer to the Python branch status, rather than the release numbers themselves: https://devguide.python.org/#status-of-python-branches

> the month after ``Python`` and would effectively only support one

Reviewer 6: (Note: referring to Python feature releases as major releases is technically incorrect, as per https://devguide.python.org/devcycle/#development-cycle, so this PEP is technically correct in calling CPython's feature releases minor releases, and its bugfix releases patch releases. But there's a difference between the terminology the docs say folks *should* be using, and what they actually use)

> the month after ``Python`` and would effectively only support one

Author: There is now a key at the top and given that as @charris says, the `major.minor.micro` terminology is common in the scipy community, I am inclined to leave the wording as it is.

version of ``Python`` that has an install base (and then the window
is 42).  This buffer also makes us resilient to minor fluctuations /
delays in the ``Python`` release schedule.

Because the threshold for dropping a version of ``Python`` is based
on historical release dates and a project's plans, the decision to drop
support for a given version of ``Python`` can be made very early in
the release process.

There will be some unavoidable mismatch in supported version of
``Python`` between packages immediately after a version of
``Python`` ages out.  However this should not last longer than one
release cycle of each of the projects, and when a given project
does a minor or major release, it is guaranteed that there will be a
stable release of all other projects that support the set of
``Python`` the new release will support.

> If there is a major version bump in Python this policy will have to be

Author: Completely re-worded this as this was about a Python 4, not a change in how often Python releases.

re-addressed as we can not predict how drastic the changes will be and
how the community will react.

Implementation
--------------

We suggest that all projects adopt the following language into their
development guidelines:

   - support minor versions of ``Python`` initially released
     42 months prior to our planned release date
   - always support at least the 2 latest minor versions of ``Python``
   - support minor versions of ``numpy`` initially released in the 24
     months prior to our planned release date or oldest that supports the
     minimum Python version (whichever is higher)

>    The minimum supported version of ``Python`` will be set to

Author: Feels different as this is a mechanics of enforcing the supported versions, where are the bullets are the rules to determine the supporting version.

   ``python_requires`` in ``setup`` and all supported versions of
   Python will be in the test matrix and have binary artifacts built
   for releases.

   We will bump the minimum Python and NumPy versions as we can on
   every minor and major release, but never on a patch release.

> For other dependencies, adopt similar time windows of 24 months or

Author: done.

shorter.

Backward compatibility
----------------------

No issues other than the intentional dropping of old version of
``Python`` and dependencies.

Alternatives
------------

Ad-Hoc
~~~~~~

Projects could on every release evaluate if they want to increase
the minimum version of Python supported.  While this is a notionally
simple policy, it makes it hard for downstream users to predict what
the future minimum versions will be.  As there is no objective threshold
to when the minimum version should be dropped, it is easy for these
discussions to devolve into [bike shedding](https://en.wikipedia.org/wiki/Wikipedia:Avoid_Parkinson%27s_bicycle-shed_effect) and acrimony.

All Python Software Foundation supported versions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> This is a very clear and conservative approach.  However, it means that

Reviewer 4: The CPython supported versions of Python are listed in the Python Developers Guide and the Python PEPs. Supporting these are a very clear...

there is 4 year lag between when new language features come into the
language and when the projects are able to use them.  Additionally,
for projects that have a significant component of compiled extensions
this requires building many binary artifacts for each release.

For the case of NumPy, many projects carry workarounds to bugs that
are fixed in subsequent versions of NumPy.  Being proactive about
increasing the minimum version of NumPy will allow downstream
packages to carry fewer version-specific patches.

Default version on Linux distribution
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The policy could be to support the version of Python that ships by
default in the latest Ubuntu LTS or CentOS/RHEL release.  However, we
would still have to standardize across the community which
distribution we are following.

By following the versions supported by major Linux distributions, we
are giving up technical control of our projects to external
organizations that may have different motivations and concerns than we
do.

N minor versions of Python
~~~~~~~~~~~~~~~~~~~~~~~~~~

Given the current release cadence of the Python, the proposed time

> (42 months) is roughly equivalent to "the last two" Python minor

Reviewer 4: minor and major is a bit confusing when it comes to Python. Perhaps define up front minor = x.x versions and dot releases (x.x.x) and major = x

Author: Added a section at the top defining these terms.

==================================================================================
NEP 29 — Recommend Python and Numpy version support as a community policy standard
==================================================================================

:Author: Thomas A Caswell <tcaswell@gmail.com>, Andreas Mueller, Brian Granger, Madicken Munk, Ralf Gommers, Matt Haberland <mhaberla@calpoly.edu>, Matthias Bussonnier <bussonniermatthias@gmail.com>, Stefan van der Walt
:Status: Draft
:Type: Informational Track
:Created: 2019-07-13

Abstract
--------

This NEP recommends and encourages all projects across the Scientific Python ecosystem to adopt a common "time window-based" policy for support of Python and NumPy versions. Standardizing a recommendation for project support of minimum Python and NumPy versions will improve downstream project planning.

> based policy for increasing the minimum version of Python and NumPy

Reviewer 4: ```suggestion
```

> that downstream projects support.  By standardizing this policy

Reviewer 4: Delete from here to end of paragraph.

across the community we will make it easier for downstream projects to
plan.

This is an unusual NEP in that it offers recommendations for
community-wide policy and not for changes to NumPy itself.  Since a
common place for SPEEPs (Scientific Python Ecosystem Enhancement
Proposals) does not exist and given NumPy's central role in the
ecosystem, a NEP provides a visible place to document the proposed
policy.

This NEP is being put forward by maintainers of Matplotlib, scikit-learn,
IPython, Jupyter, yt, SciPy, NumPy, and scikit-image.

Detailed description
--------------------

For the purposes of this NEP we assume semantic versioning and define:

> *major version*

Reviewer 4: Nicely improved.

IPython, Jupyter, yt, SciPy, NumPy, and scikit-image.

Detailed description
--------------------

For the purposes of this NEP we assume semantic versioning and define:

*major version*
   A release that change the first number (e.g. X.0.0)

*minor version*
   A release that changes the second number (e.g x.Y.0)

*patch version*
   A release that changes the third number (e.g. x.y.Z)

When a project creates a new major or minor version, we recommend that
the project should support all minor versions of ``Python`` introduced
and released in the prior 42 months ~~from their anticipated release
date~~ and all minor versions of NumPy released in the prior 24
months.

The diagram::

       Jan 16      Jan 17      Jan 18      Jan 19      Jan 20
       |           |           |           |           |
  +++++|+++++++++++|+++++++++++|+++++++++++|+++++++++++|++++++++++++
   |              |                  |               |
   py 3.5.0       py 3.6.0           py 3.7.0        py 3.8.0
  |-----------------------------------------> Feb19
            |-----------------------------------------> Dec19
                      |-----------------------------------------> Nov20

> shows the 42 month support windows for ``Python``.  A project with a

Reviewer 7: "``Python``" here, "Python" elsewhere.

major or minor version release in Feb19 should support py35 and newer,
a project with a major or minor version release in Dec19 should
support py36 and newer, and a project with a major or minor version
release in Nov20 should support py37 and newer.

The current Python release cadence is 18 months so a 42 month window
ensures that there will always be at least two versions of ``Python``
in the window.  By padding the window by 6 months from the anticipated
``Python`` cadence we avoid the edge cases where a project releases
the month after ``Python`` and would effectively only support one
version of ``Python`` that has an installed base.
This six month buffer provides resilience to minor fluctuations /
delays in the ``Python`` release schedule.

Because ``Python`` version support is based on historical release
dates, a 42 month time window, and a project's plans, a project can
decide to drop a given version of ``Python`` very early in the release
process.

While there will be some unavoidable mismatch in supported version of

> ``Python`` between if a project release occurs immediately after a

Reviewer 7: between....projects?

version of ``Python`` ages out.  This should not last longer than one
release cycle of each of the projects, and when a given project does a
minor or major release, it is guaranteed that there will be a stable

> release of all other projects that support the set of ``Python`` the

Reviewer 7: the set of Python versions?

when the projects are able to use them.  Additionally, for projects
that have a significant component of compiled extensions this requires
building many binary artifacts for each release.

For the case of NumPy, many projects carry workarounds to bugs that
are fixed in subsequent versions of NumPy.  Being proactive about
increasing the minimum version of NumPy will allow downstream
packages to carry fewer version-specific patches.

Default version on Linux distribution
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The policy could be to support the version of Python that ships by
default in the latest Ubuntu LTS or CentOS/RHEL release.  However, we
would still have to standardize across the community which
distribution we are following.

By following the versions supported by major Linux distributions, we
are giving up technical control of our projects to external
organizations that may have different motivations and concerns than we
do.

N minor versions of Python
~~~~~~~~~~~~~~~~~~~~~~~~~~

Given the current release cadence of the Python, the proposed time
(42 months) is roughly equivalent to "the last two" Python minor

> versions.  However, if Python changes their release cadence, any rule

Reviewer 7: ```suggestion
versions.  However, if Python changes their release cadence substantially, any rule
```

> based on the number of minor releases will need to be changed.

Reviewer 7: ```suggestion
based solely on the number of minor releases may need to be changed to remain sensible.
```

> 

Reviewer 7: A rule based on the number of minor releases wouldn't _inherently_ have to change if the release cadence were to change by epsilon. Only if the change in release cadence is sufficiently large might the rule might have to change, and then only if we require that the rule remain sensible : )
I think that the real point is that the rule shouldn't be based _solely_ on minor versions **or** _solely_ on time; a composite rule is more robust to changes in release cadence.

Reviewer 6: I think it's also reasonable for the Scientific Python community to ask Python core to use "Impact on the Scientific Python community's standard support policy" as one of our points of consideration if we're considering adjusted our release cadence (which we are actually doing at the moment - we'd like to align it with the calendar year, which means either an every-year cadence, or an every-other-year one).

So if you want to consider that potential impact in advance and design a rule around it, the two scenarios to consider would be:

1. What if the current 18 months became 12 months?
2. What if the current 18 months became 24 months?

(If attempting to design such a rule results in a conclusion of "That cadence would be awful for us", then that's important feedback for us to hear *before* we make any decisions)

Reviewer 2: > What if the current 18 months became 12 months?

This would probably be fine, and we could leave this policy unchanged at 42 months - that would make packages oscillate between supporting 3 and 4 Python releases, rather than 2-3. Taking a year off our support would be too aggressive imho - we'd get many more users who we're then forcing to upgrade their Python install (can be painful).

> What if the current 18 months became 24 months?

that would be awful

Reviewer 6: Note that there's more to the 24 month proposal than slowing down the breaking releases - there are 12 months of backwards compatible incremental feature releases, so both proposals reduce the  "no CPython feature releases" gap to at most 12 months. It's just one of them also aims to reduce the frequency of "rebuild the world" events.

Time window on the X.Y.1 Python release
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As the first bug fix release is typically a few months after the
initial release, you can achieve the same effect by using a large delay

> from the X.Y.0 release which seems simpler to explain.

Reviewer 7: I still have a hard time understanding this one. I'm assuming "a large delay from the X.Y.0 release" refers to the 6 month buffer on top of the 2 x 18 month time-based rule? 

Let's at least keep the language consistent with the definitions above. Consider: 
"As the first **patch version** is typically released within a few months of the corresponding **minor version**,  a rule based on the first patch release is effectively similar to the proposed rule, given the proposed rule's six-month buffer."

(This isn't strictly true if the release cadence were to change, right?)

One option is to remove this alternative entirely, unless you think it would be missed. If so, do you need to mention rules based on X.Y.2? X.Y.3? (Probably not.)

Author: Tried re-writing this again, if it is still not clear we can axe it!
